### DISCUSSION TITLE
Rounded corners on charts

### DISCUSSION THREAD

**Discussion by perNyfelt** [eclipse-birt/birt#1544] [Q&A] [answered]
Is there some way to produce bar charts, grid borders and table borders with rounded corners?

**Comment by speckyspooky** [comment]
The question would be which output you wants to change. But at the end there is no default way to do it with default settings.

In the past I changed the HTML-preview with enhanced JS to add special CSS and so I got rounded corners.
But the CSS instructions won't be interpreted through the output emitter of PDF or MS Word because it is special CSS.
Yes PDF would be able to draw rounded corners but there is no interpretation of the advanced CSS instruction implemented.
My HTML-manipulation works only due to the browser interpretation.

The charts could be changed on HTML-level through the same manipulation (if the chart create a SVG) but I think this won't help you
because I think you will need it on printable output like PDF.

**Comment by perNyfelt** [comment]
Thank you! The output is PDF and the charts create SVG's (at least they are SVG's when i generate the reports in HTML format).

I would like to help adding this feature. Do you know if anyone is working on this? If not would you mind giving me me a few more pointers on what implementing support for this would entail (I have not contributed any code to Birt before)?

**Comment by hvbtup** [comment, marked as answer]
AFAIK if you have enabled SVG charts, then these are converted to PDF vector graphic instructions by Apache Batik when BIRT creates a PDF.
And if the chart is converted to a bitmap for other output formats, then rendering the SVG chart as a bitmap is handled by Batik, too.

So, implementing rounded corners inside charts would be possible relatively easy by changing the way the SVG is generated.
You would be able to do this "simply" by manipulating the Java sources for the chart plugin, I think.

This is different from implementing rounded corners for grids, tables etc.
For these items, the various border properties are translated into emitter-specific instructions, so they are different for PDF, HTML, Word, Excel output. 
Thomas (@speckyspooky) recently added diagonal lines to the border properties.
I think what you would need to do for grids, tables, cells, would be very similar to what he did for the diagonal lines.

Besides taking care of rendering rounded corners, you would need an option to configure them.
This would include extending the ROM and extending the UI dialogs for specifying them.

Personally, I am not sure if this is really a good idea. AFAIK, BIRT stores various properties for each and every layout item generated.
Say, we have 60 properties now, for example. Adding eg. 4 more properties (a radius each of the four corners) increases this size.
So this means we have ~ 7% more properties. This could result in a few percent more memory consumption and a slightly increased CPU consumption. Not sure if this is worth it for a feature that most users would consider "nice to have, but I don't need it".

What I would do:
1) For the moment, take a fixed radius and try to modify the sources of the chart plugin in such a way that it produces a SVG with rounded corners.
2) Check if these rounded corners work with the most important emitters (PDF, HTML, Word, Excel) or at least where you need them.
3) Create a report where you set a user-property for the corner-radius (the same for all 4 corners) at one specific case.
4) Adapt the code from 1) to override the standard radius with the value from the user property, if given.
5) Repeat the tests from 2) and see if the overriding works.
6) Change the constant radius to 0 (or null) again, so that without specifying the user property, the report behaves as before (no rounded corners)

As far as you are concerned, you may stop here, because you can use the user property to create rounded corners.
You could then contribute your code to the standard as a PR.

We would then need to discuss 
1) if we take it as is, or if we (you) should extend the ROM, the property editors etc to make this more user-friendly (but as I said before, this might result in increased ressource consumption at runtime).
2) if we should try to implement rounded corners for grid/table/cells as well (taking Thomas' changes for diagonal borders as a template).

**Reply by perNyfelt** [reply]
Great thank you! Very good explanation, I will give it a try and get back with a PR based on user properties if I can get it to work.
